Add feature to enable Time based DND Mode

manavmehta opened the pull request:
**What's this PR do?**
Fixes: #561
User can now select time for which DND will be enabled.
Toast notification (4 sec) will be shown for clock time when DND would go off and also when it actually goes off.

**Screenshots?**
![Screenshot 2020-03-11 at 12 59 19 AM](https://user-images.githubusercontent.com/31139861/76351500-90150500-6333-11ea-94d1-88597429472b.png)
![Screenshot 2020-03-11 at 12 58 29 AM](https://user-images.githubusercontent.com/31139861/76351507-91dec880-6333-11ea-9169-336ac39d8e30.png)

**You have tested this PR on:**
[✓ ] macOS Catalina 10.15.4

manavmehta commented:
@timabbott @akashnimare @abhigyank 
Can you please review and suggest !?!

abhigyank commented on app/renderer/css/main.css:
```diff
@@ -82,6 +82,42 @@ body {
 
+/* DND new start */
+
+.dnd-tooltip {
```
Could rename this as `dnd-menu` .

abhigyank commented on app/renderer/js/main.ts:
```diff
@@ -400,8 +400,17 @@ class ServerManagerView {
+			}
+			if (document.querySelector('.dnd-tooltip') === null) {
+				DNDUtil.makeView();
+				// return; // linter error
```
No point of this, we can remove this.

abhigyank commented on app/renderer/js/utils/dnd-util.ts:
```diff
@@ -46,3 +54,104 @@ export function toggle(): Toggle {
+};
+
+const configureDND = (opt: HTMLLIElement, dndMenu: HTMLUListElement): void => { // Configures the DND as per user's choice
+	let dndOffTime;
```
The indentation seems off here.

manavmehta replied:
@akashnimare can you check if the same problem is there for you too as @abhigyank 's system showed no indent.
But mine shows correctly.
P.S. Linter also doesn't throw any error

abhigyank replied:
Looks fine for me now.
Probably some loading error on Github,

abhigyank commented on app/renderer/js/utils/dnd-util.ts:
```diff
@@ -46,3 +54,104 @@ export function toggle(): Toggle {
+	const actionContainer = document.querySelector('#actions-container');
+	dndMenu.style.height = '75px';
+	optionElement.className = 'dnd-options-btn';
+	if (opt.id === 'thirty_min') {
```
Instead of an if-else block, it would be preferable to go for a switch-case.

abhigyank commented on app/renderer/js/utils/dnd-util.ts:
```diff
@@ -46,3 +54,104 @@ export function toggle(): Toggle {
+	ipcRenderer.send('forward-message', 'toggle-dnd', state.dnd, state.newSettings);
+};
+
+let job: any = null; // Type annotation
```
You should move this to the top, since its also used in functions above.

abhigyank commented:
Left some comments. I like how it works smoothly for both the themes.

manavmehta commented:
@abhigyank Thanks. I have implemented your suggestions and toast is also done now. So it's a completed PR from my end. 
@akashnimare @timabbott Please review. Suggestions welcomed :)

abhigyank commented on app/renderer/js/main.ts:
```diff
@@ -400,8 +400,16 @@ class ServerManagerView {
+				const dndUtil = DNDUtil.toggle();
+				ipcRenderer.send('forward-message', 'toggle-dnd', dndUtil.dnd, dndUtil.newSettings);
+			}
+			if (document.querySelector('.dnd-menu') === null) {
```
Should be an `else if` block.

akashnimare replied:
@manavmehta can you make the above changes?

manavmehta replied:
Done

abhigyank commented:
The working is fine, it is working as expected for me. The overall logic implementation looks correct to me. 
@akashnimare Can you do a code quality check? It is then good to merge.

akashnimare commented on app/renderer/js/utils/dnd-util.ts:
```diff
@@ -37,12 +43,128 @@ export function toggle(): Toggle {
+	if (dndExist.length === 0) {
+		const dndMenu = document.createElement('ul');
+		dndMenu.className = 'dnd-menu';
+		const dndOptions = [['30 Minutes', 'thirty_min'],
```
@manavmehta @abhigyank this can be improved I think.

manavmehta replied:
@akashnimare you mean creating the menu ?

abhigyank replied:
@manavmehta I think it'd be better to use a `map` here - 
```js
let dndOptions = new Map();
dndOptions.set('thirty_minutes', '30 Minutes');
dndOptions.set('one_hour', '1 Hour');
...
```
And in the below iteration it'd be good to use `forEach` - 
```js
dndOptions.forEach((value, key) => {
  ...
})
```

manavmehta replied:
used an object as akash suggested
if you find map more useful I can do the necessary change

akashnimare commented:
@manavmehta can you refactor the PR a bit (there are some places where you can improve the code quality + add comments etc)?

manavmehta reviewed (commented):
@akashnimare any specifics for refactoring - any function

> @manavmehta can you refactor the PR a bit (there are some places where you can improve the code quality + add comments etc)?

Any specific block for refactoring and improving code quality? I can surely implement if you mention.

manavmehta committed:
dnd: Add feature to enable time based DND Mode

Create a menu to allow selecting time for which DND will be enabled.
DND will go off at the scheduled time and a toast will be shown, both while setting the DND, thus conveying the DND Off time and again at the time when DND goes off.

'Until I resume' option is time independent, trivially.

Fixes #561

akashnimare commented:
I have sent you a few things which you can improve in the PM.

manavmehta commented:
@akashnimare @abhigyank implemented. Please review.
Also if you may see if map would be better than object in makeView()

akashnimare commented:
This looks better I think. @andersk quick review?

abhigyank commented:
Working LGTM. Refactoring has improved the code. Can be merged after @andersk  does a review.

manavmehta commented:
@andersk ping

abhigyank commented:
@manavmehta  Rebase this once you're free. We can merge this @akashnimare .

manavmehta commented:
@abhigyank Rebased. Tested again on macOS.

akashnimare commented on app/renderer/css/main.css:
```diff
@@ -82,6 +82,59 @@ body {
+/* DND Menu */
+
+/* stylelint-disable-next-line */
+@keyframes fading{0%{opacity:0}20%{opacity:1}80%{opacity:1}100%{opacity:0}}
```
@manavmehta what happened to linting here?

manavmehta replied:
@akashnimare 
Toast fading out. I think I borrowed it from some module.
You want me to spread it ?

akashnimare replied:
No no, just run the linting that's it.

manavmehta replied:
There will be linting errors if stylelint is not disabled for that line. It requires spreading. With the exception, it gives no errors. Checked

abhigyank replied:
@manavmehta I think it should be something like this - 
```css
@keyframes fading {
    0% {
        opacity: 0
    }
    20% {
        opacity: 1
    }
    80% {
        opacity: 1
    }
    100% {
        opacity: 0
    }
}
```

manavmehta replied:
I had committed the changes, just waiting for either of you to suggest either to go for these or not. Pushed now. Thanks

akashnimare commented on app/renderer/css/main.css:
```diff
@@ -82,6 +82,59 @@ body {
+    padding: 15px 30px  15px 30px;
+    color: rgb(0, 0, 0);
+    animation: fading 4s;
+    font-size: 13.5px;
```
There are two font-sizes in this class. Please remove one.

abhigyank commented on app/renderer/js/utils/dnd-util.ts:
```diff
@@ -46,3 +55,125 @@ export function toggle(): Toggle {
+		actionsContainer.append(dndMenu);
+
+		const dndMenuFocusOut = (event: Event) => {
+			console.log(event);
```
Remove this log statement.

abhigyank reviewed (commented):
`dndswitchOff` would probably be better named as `dndEndTime` ?

manavmehta committed:
Refactoring:

* Better documentation
* Use better methods like `append` instead of `innerHTML`
* Better function names and parameters

abhigyank commented on app/renderer/js/main.ts:
```diff
@@ -403,8 +403,15 @@ class ServerManagerView {
+				DNDUtil.makeMenu();
+			} else {
+				document.querySelector('.dnd-menu').remove();
+			}
```
Can you put some comments describing the scenarios of these if-else conditions?

abhigyank commented on app/renderer/js/utils/dnd-util.ts:
```diff
@@ -46,3 +55,126 @@ export function toggle(): Toggle {
+/*	Configure DND as per user's selected option and pops a toast showing dnd switch off time. */
+const configureDND = (dndTime: string, dndMenu: HTMLUListElement): void => {
+	let dndOffTime;
+	dndOffTime = new Date();
```
Why these two as separate lines? Can be in one line I think.
